File: app/src/main/java/sugtao4423/lod/tweetlistview/TweetListUserAdapter.java
package sugtao4423.lod.tweetlistview;

import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.support.v7.widget.RecyclerView;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.HorizontalScrollView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.loopj.android.image.SmartImageView;

import java.text.NumberFormat;
import java.util.ArrayList;

import sugtao4423.lod.App;
import sugtao4423.lod.R;
import sugtao4423.lod.userpage_fragment.UserPage;
import twitter4j.PagableResponseList;
import twitter4j.User;

public class TweetListUserAdapter extends RecyclerView.Adapter<TweetListUserAdapter.ViewHolder>{

    private LayoutInflater inflater;
    private ArrayList<User> data;
    private Context context;
    private App app;
    private Handler handler;

    public TweetListUserAdapter(Context context){
        this.inflater = LayoutInflater.from(context);
        this.data = new ArrayList<User>();
        this.context = context;
        this.app = (App)context.getApplicationContext();
        this.handler = new Handler();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int position){
        return new ViewHolder(inflater.inflate(R.layout.list_item_tweet, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position){
        if(!(data != null && data.size() > position && data.get(position) != null)){
            return;
        }

        holder.rt_icon.setVisibility(View.GONE);
        holder.rt_sn.setVisibility(View.GONE);
        holder.tweetImagesScroll.setVisibility(View.GONE);

        final User item = data.get(position);

        if(item.isProtected()){
            holder.protect.setTypeface(app.getFontAwesomeTypeface());
            holder.protect.setTextSize(TypedValue.COMPLEX_UNIT_SP, app.getOptions().getUserNameFontSize() - 3);
            holder.protect.setVisibility(View.VISIBLE);
        }else{
            holder.protect.setVisibility(View.GONE);
        }

        holder.icon.setImageUrl(item.getBiggerProfileImageURLHttps(), null, R.drawable.icon_loading);
        holder.name_sn.setText(item.getName() + " - @" + item.getScreenName());
        holder.content.setText(item.getDescription());
        String userCountsText = context.getString(R.string.param_user_count_detail,
                numberFormat(item.getStatusesCount()),
                numberFormat(item.getFavouritesCount()),
                numberFormat(item.getFriendsCount()),
                numberFormat(item.getFollowersCount())
        );
        holder.date.setText(userCountsText);

        holder.name_sn.setTextSize(TypedValue.COMPLEX_UNIT_SP, app.getOptions().getUserNameFontSize());
        holder.content.setTextSize(TypedValue.COMPLEX_UNIT_SP, app.getOptions().getContentFontSize());
        holder.date.setTextSize(TypedValue.COMPLEX_UNIT_SP, app.getOptions().getDateFontSize());

        holder.v.setOnClickListener(new OnClickListener(){

            @Override
            public void onClick(View v){
                Intent i = new Intent(context, UserPage.class);
                i.putExtra(UserPage.INTENT_EXTRA_KEY_USER_SCREEN_NAME, item.getScreenName());
                context.startActivity(i);
            }
        });
    }

    public String numberFormat(int num){
        return NumberFormat.getInstance().format(num);
    }

    @Override
    public int getItemCount(){
        if(data != null){
            return data.size();
        }else{
            return 0;
        }
    }

    public void addAll(final PagableResponseList<User> users){
        final int pos = data.size();
        data.addAll(users);
        handler.post(new Runnable(){

            @Override
            public void run(){
                notifyItemRangeInserted(pos, users.size());
            }
        });
    }

    public void clear(){
        final int size = data.size();
        data.clear();
        handler.post(new Runnable(){

            @Override
            public void run(){
                notifyItemRangeRemoved(0, size);
            }
        });
    }

    class ViewHolder extends RecyclerView.ViewHolder{

        View v;
        SmartImageView icon, rt_icon;
        TextView name_sn, content, date, rt_sn, protect;
        HorizontalScrollView tweetImagesScroll;
        LinearLayout tweetImagesLayout;

        public ViewHolder(View itemView){
            super(itemView);
            icon = (SmartImageView)itemView.findViewById(R.id.icon);
            rt_icon = (SmartImageView)itemView.findViewById(R.id.RetweetedUserIcon);
            name_sn = (TextView)itemView.findViewById(R.id.name_screenName);
            content = (TextView)itemView.findViewById(R.id.tweetText);
            date = (TextView)itemView.findViewById(R.id.tweet_date);
            rt_sn = (TextView)itemView.findViewById(R.id.RetweetedUserScreenName);
            protect = (TextView)itemView.findViewById(R.id.UserProtected);
            tweetImagesScroll = (HorizontalScrollView)itemView.findViewById(R.id.tweet_images_scroll);
            tweetImagesLayout = (LinearLayout)itemView.findViewById(R.id.tweet_images_layout);
            v = itemView;
        }
    }

}

File: app/src/main/java/sugtao4423/lod/tweetlistview/TweetListUserAdapter.kt
package sugtao4423.lod.tweetlistview

import android.content.Context
import android.content.Intent
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.HorizontalScrollView
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import sugtao4423.lod.App
import sugtao4423.lod.R
import sugtao4423.lod.userpage_fragment.UserPage
import twitter4j.PagableResponseList
import twitter4j.User
import java.text.NumberFormat

class TweetListUserAdapter(private val context: Context) : RecyclerView.Adapter<TweetListUserAdapter.ViewHolder>() {

    private val inflater = LayoutInflater.from(context)
    private val data = ArrayList<User>()
    private val app = context.applicationContext as App

    override fun onCreateViewHolder(viewGroup: ViewGroup, position: Int): ViewHolder {
        return ViewHolder(inflater.inflate(R.layout.list_item_tweet, viewGroup, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (data.size <= position) {
            return
        }
        val item = data[position]

        holder.rtIcon.visibility = View.GONE
        holder.rtSn.visibility = View.GONE
        holder.tweetImagesScroll.visibility = View.GONE
        holder.tweetImagesLayout.visibility = View.GONE

        if (item.isProtected) {
            holder.protect.apply {
                visibility = View.VISIBLE
                typeface = app.fontAwesomeTypeface
                setTextSize(TypedValue.COMPLEX_UNIT_SP, app.prefRepository.userNameFontSize - 3)
            }
        } else {
            holder.protect.visibility = View.GONE
        }

        Glide.with(context).load(item.biggerProfileImageURLHttps).placeholder(R.drawable.icon_loading).into(holder.icon)
        holder.nameSn.text = item.name + " - @" + item.screenName
        holder.content.text = item.description
        val userCountsText = context.getString(R.string.param_user_count_detail,
                numberFormat(item.statusesCount),
                numberFormat(item.favouritesCount),
                numberFormat(item.friendsCount),
                numberFormat(item.followersCount)
        )
        holder.date.text = userCountsText

        holder.nameSn.setTextSize(TypedValue.COMPLEX_UNIT_SP, app.prefRepository.userNameFontSize)
        holder.content.setTextSize(TypedValue.COMPLEX_UNIT_SP, app.prefRepository.contentFontSize)
        holder.date.setTextSize(TypedValue.COMPLEX_UNIT_SP, app.prefRepository.dateFontSize)

        holder.itemView.setOnClickListener {
            val i = Intent(context, UserPage::class.java)
            i.putExtra(UserPage.INTENT_EXTRA_KEY_USER_SCREEN_NAME, item.screenName)
            context.startActivity(i)
        }
    }

    private fun numberFormat(num: Int): String {
        return NumberFormat.getInstance().format(num)
    }

    override fun getItemCount(): Int {
        return data.size
    }

    fun addAll(users: PagableResponseList<User>) {
        val pos = data.size
        data.addAll(users)
        notifyItemRangeInserted(pos, users.size)
    }

    fun clear() {
        val size = data.size
        data.clear()
        notifyItemRangeRemoved(0, size)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val icon: ImageView = itemView.findViewById(R.id.tweetIcon)
        val rtIcon: ImageView = itemView.findViewById(R.id.retweetedUserIcon)
        val nameSn: TextView = itemView.findViewById(R.id.tweetNameScreenName)
        val content: TextView = itemView.findViewById(R.id.tweetText)
        val date: TextView = itemView.findViewById(R.id.tweetDate)
        val rtSn: TextView = itemView.findViewById(R.id.retweetedUserScreenName)
        val protect: TextView = itemView.findViewById(R.id.tweetUserProtected)
        val tweetImagesScroll: HorizontalScrollView = itemView.findViewById(R.id.tweetImagesScroll)
        val tweetImagesLayout: LinearLayout = itemView.findViewById(R.id.tweetImagesLayout)
    }

}
